src/network/C4HTTP.cpp

```diff
@@ -0,0 +1,307 @@
+/*
+ * OpenClonk, http://www.openclonk.org
+ *
+ * Copyright (c) 2017, The OpenClonk Team and contributors
+ *
+ * Distributed under the terms of the ISC license; see accompanying file
+ * "COPYING" for details.
+ *
+ * "Clonk" is a registered trademark of Matthes Bender, used with permission.
+ * See accompanying file "TRADEMARK" for details.
+ *
+ * To redistribute this file separately, substitute the full license texts
+ * for the above references.
+ */
+#include "C4Include.h"
+#include "network/C4HTTP.h"
+#include "network/C4InteractiveThread.h"
+#include "C4Version.h"
+
+#include <regex>
+
+#define CURL_STRICTER
+#include <curl/curl.h>
+
+// *** C4HTTPClient
+
+C4HTTPClient::C4HTTPClient()
+{
+}
+
+C4HTTPClient::~C4HTTPClient()
+{
+	Cancel(nullptr);
+	if (MultiHandle)
+		curl_multi_cleanup(MultiHandle);
+#ifdef STDSCHEDULER_USE_EVENTS
+	if (Event != nullptr)
+		WSACloseEvent(Event);
+#endif
+}
+
+bool C4HTTPClient::Init()
+{
+	MultiHandle = curl_multi_init();
+	if (!MultiHandle) return false;
+#ifdef STDSCHEDULER_USE_EVENTS
+	if ((Event = WSACreateEvent()) == WSA_INVALID_EVENT)
+	{
+		SetError("could not create socket event");
+		curl_multi_cleanup(MultiHandle);
+		return false;
+	}
+#endif
+	curl_multi_setopt(MultiHandle, CURLMOPT_SOCKETFUNCTION, &C4HTTPClient::SSocketCallback);
+	curl_multi_setopt(MultiHandle, CURLMOPT_SOCKETDATA, this);
+	return true;
+}
+
+bool C4HTTPClient::Execute(int iMaxTime, pollfd *readyfd)
+{
+	int running;
+#ifdef STDSCHEDULER_USE_EVENTS
+	// On Windows, StdScheduler doesn't inform us about which fd triggered the
+	// event, so we have to check manually.
+	if (WaitForSingleObject(Event, 0) == WAIT_OBJECT_0)
+	{
+		for (const auto& kv : sockets)
```

Comment:
This is going to crash on Windows as the socket callback called by `curl_multi_socket_action` may erase the socket out of the map, and changing the map in the body of a ranged for loop isn't allowed.